### PR TITLE
net/dns, paths, util/winutil: change net/dns/windowsManager NRPT mana…

### PR DESCRIPTION
**Pull request by dblohm7**
…gement to support more than 50 domains.

AFAICT this isn't documented on MSDN, but based on the issue referenced below,
NRPT rules are not working when a rule specifies > 50 domains.

This patch modifies our NRPT rule generator to split the list of domains
into chunks as necessary, and write a separate rule for each chunk.

For compatibility reasons, we continue to use the hard-coded rule ID, but
as additional rules are required, we generate new GUIDs. Those GUIDs are
stored under the Tailscale registry path so that we know which rules are ours.

I made some changes to winutils to add additional helper functions in support
of both the code and its test: I added additional registry accessors, and also
moved some token accessors from paths to util/winutil.

Fixes https://github.com/tailscale/coral/issues/63

Signed-off-by: Aaron Klotz <aaron@tailscale.com>

### CODE REVIEW

File: net/dns/manager_windows.go [hunk 1]
```diff
@@ -104,31 +134,95 @@ func delValue(key registry.Key, name string) error {
...
+	dr := len(domains) % nrptMaxDomainsPerRule
+
+	domainRulesLen := dq
+	if dr > 0 {
+		domainRulesLen++
+	}
+
+	nrptRuleIDs := loadRuleSubkeyNames()
+	for len(nrptRuleIDs) < domainRulesLen {
+		guid, err := windows.GenerateGUID()
```
Comment (sailorfrag) [inline net/dns/manager_windows.go:165]: So when I was looking at GPO examples online, it looks like the registry key doesn't actually have to be a GUID, that's just a convention.

I've seen other examples that amounted to the equivalent of Tailscale1, Tailscale2, Tailscale3, etc.

File: net/dns/manager_windows.go [hunk 2]
```diff
@@ -90,10 +109,21 @@ func (m windowsManager) ifPath(basePath string) string {
...
 		return err
 	}
+	return winutil.DeleteRegValue(nrptRuleIDValueName)
+}
+
+func delRuleKeys(nrptRuleIDs []string) error {
+	for _, rid := range nrptRuleIDs {
+		keyName := nrptBase + rid
+		if err := registry.DeleteKey(registry.LOCAL_MACHINE, keyName); err != nil && err != registry.ErrNotExist {
+			return err
```
Comment (sailorfrag) [inline net/dns/manager_windows.go:124]: Would it be better to log and continue?

File: net/dns/manager_windows.go [hunk 3]
```diff
@@ -59,7 +78,7 @@ func NewOSConfigurator(logf logger.Logf, interfaceName string) (OSConfigurator,
 	// boot up. The bootstrap resolver logic will save us, but it
 	// slows down start-up a bunch.
 	if ret.nrptWorks {
-		ret.delKey(nrptBase)
+		delAllRuleKeys()
```
Comment (sailorfrag) [inline net/dns/manager_windows.go:81]: Log the error?

File: net/dns/manager_windows.go [hunk 4]
```diff
@@ -104,31 +134,95 @@ func delValue(key registry.Key, name string) error {
...
+		guid, err := windows.GenerateGUID()
+		if err != nil {
+			return err
+		}
+		nrptRuleIDs = append(nrptRuleIDs, guid.String())
+	}
+
+	// Remove any surplus rules that are no longer needed.
+	ruleIDsToRemove := nrptRuleIDs[domainRulesLen:]
+	delRuleKeys(ruleIDsToRemove)
```
Comment (sailorfrag) [inline net/dns/manager_windows.go:174]: log errors

File: net/dns/manager_windows.go [hunk 5]
```diff
@@ -104,31 +134,95 @@ func delValue(key registry.Key, name string) error {
...
+
+	// We need to save the list of rule IDs to our Tailscale registry key so that
+	// we know which rules are ours during subsequent modifications to NRPT rules.
+
+	// Omit nrptSingleRuleID at index 0 when writing our list of rule IDs to the
```
Comment (sailorfrag) [inline net/dns/manager_windows.go:179]: Seems like it would be simpler just to include `nrptSingleRuleID`?

**Review by sailorfrag — APPROVED** [review]
Also https://github.com/tailscale/tailscale/issues/4607#issuecomment-1130586168 indicates that maybe there's some kind of thing we should be doing to kick DNSClient into reloading its config.

File: net/dns/manager_windows.go (hunk 1 above)
Comment (dblohm7) [reply net/dns/manager_windows.go:165]: Per [this document](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-gpnrpt/2d34f260-1e9e-4a52-ac91-2056dfd29702) it is a GUID.

File: net/dns/manager_windows.go (hunk 2 above)
Comment (dblohm7) [reply net/dns/manager_windows.go:124]: Yeah, I'll change this.

File: net/dns/manager_windows.go (hunk 3 above)
Comment (dblohm7) [reply net/dns/manager_windows.go:81]: I updated the implementation to log.

File: net/dns/manager_windows.go (hunk 5 above)
Comment (dblohm7) [reply net/dns/manager_windows.go:179]: Yeah, I'm revising this.

**Pull request merged**